File: 0/049/GroupAnagrams.cs
/*
https://leetcode.com/problems/group-anagrams/description/?envType=daily-question&envId=2024-02-06
*/

public class Solution {
    public IList<IList<string>> GroupAnagrams(string[] strs) {
        int sz=strs.Length;
        Dictionary<string, List<int>>mp=new Dictionary<string, List<int>>();
        for(int i=0; i<sz; ++i){
            string tmp=strs[i];
            char[] tmpchararr=tmp.ToCharArray();
            Array.Sort(tmpchararr);
            string sortedtmp=new string(tmpchararr);
            if(mp.ContainsKey(sortedtmp)){
                List<int>sortedtmpvals=mp[sortedtmp];
                sortedtmpvals.Add(i);
            } else{
                List<int>nwlist=new List<int>();
                nwlist.Add(i);
                mp[sortedtmp]=nwlist;
            }
        }
        IList<IList<string>>res=new List<IList<string>>();
        foreach((string key, List<int>vals) in mp){
            IList<string>curres=new List<string>();
            List<int>curmpvec=vals;
            int curentrysz=curmpvec.Count;
            for(int i=0; i<curentrysz; ++i){
                curres.Add(strs[curmpvec[i]]);
            }
            res.Add(curres);
        }
        return res;
    }
}

File: 0/049/GroupAnagrams.cpp
/*
https://leetcode.com/problems/group-anagrams/description/?envType=daily-question&envId=2024-02-06
*/

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int sz=strs.size();
        unordered_map<string, vector<int>>mp;
        for(int i=0; i<sz; ++i){
            string tmp=strs[i];
            sort(tmp.begin(), tmp.end());
            mp[tmp].push_back(i);
        }
        vector<vector<string>>res;
        for(auto &entries: mp){
            vector<string>curres;
            vector<int>curmpvec=entries.second;
            int curentrysz=curmpvec.size();
            for(int i=0; i<curentrysz; ++i){
                curres.push_back(strs[curmpvec[i]]);
            }
            res.push_back(curres);
        }
        return res;
    }
};